lighthouse-core/audits/seo/font-size.js

```diff
@@ -121,80 +120,84 @@ function nodeToTableNode(node) {
 function findStyleRuleSource(baseURL, styleDeclaration, node) {
-  if (
-    !styleDeclaration ||
+  if (!styleDeclaration ||
     styleDeclaration.type === 'Attributes' ||
     styleDeclaration.type === 'Inline'
   ) {
     return {
+      source: {type: 'url', value: baseURL},
       selector: nodeToTableNode(node),
-      source: baseURL,
     };
   }
 
   if (styleDeclaration.parentRule &&
     styleDeclaration.parentRule.origin === 'user-agent') {
     return {
+      source: {type: 'code', value: 'User Agent Stylesheet'},
       selector: styleDeclaration.parentRule.selectors.map(item => item.text).join(', '),
-      source: 'User Agent Stylesheet',
     };
   }
 
-  if (styleDeclaration.type === 'Regular' && styleDeclaration.parentRule) {
+  let selector = '';
+  if (styleDeclaration.parentRule) {
     const rule = styleDeclaration.parentRule;
-    const stylesheet = styleDeclaration.stylesheet;
-
-    if (stylesheet) {
-      let source;
-      const selector = rule.selectors.map(item => item.text).join(', ');
-
-      if (stylesheet.sourceURL) {
-        const url = new URL(stylesheet.sourceURL, baseURL);
-        const range = styleDeclaration.range;
-        source = `${url.href}`;
-
-        // !!range == has defined location in a source file (.css or .html)
-        if (range) {
-          let line = range.startLine + 1;
-          let column = range.startColumn;
-
-          // Add the startLine/startColumn of the <style> element to the range, if stylesheet
-          // is inline.
-          // Always use the rule's location if a sourceURL magic comment is
-          // present (`hasSourceURL` is true) - this makes the line/col relative to the start
-          // of the style tag, which makes them relevant when the "file" is open in DevTool's
-          // Sources panel.
-          const addHtmlLocationOffset = stylesheet.isInline && !stylesheet.hasSourceURL;
-          if (addHtmlLocationOffset) {
-            line += stylesheet.startLine;
-            // The column the stylesheet begins on is only relevant if the rule is declared on the same line.
-            if (range.startLine === 0) {
-              column += stylesheet.startColumn;
-            }
-          }
-
-          source += `:${line}:${column}`;
-        }
-      } else {
-        // dynamically injected to page
-        source = 'dynamic';
-      }
+    selector = rule.selectors.map(item => item.text).join(', ');
+  }
 
-      return {
-        selector,
-        source,
-      };
+  if (styleDeclaration.stylesheet && !styleDeclaration.stylesheet.sourceURL) {
+    // Dynamically injected into page.
+    return {
+      source: {type: 'code', value: 'dynamic'},
+      selector,
+    };
+  }
+
+  // !!range == has defined location in a source file (.css or .html)
+  if (styleDeclaration.stylesheet && styleDeclaration.range) {
+    const {range, stylesheet} = styleDeclaration;
+
+    // If a magic sourceURL comment is present, it is typically as a URL relative to the
```

Comment:
OK, so I've read this comment a few times, but I'm still not sure what, `sourceURL`, `hasSourceURL` and `canLink` represent in this saga :)

- has magic sourceURL comment === `hasSourceURL` or `!!sourceURL`?
- `canLink` === the source URL was a magic comment that points to a real file, or the source URL was not a magic comment at all and was instead automatically determined, is that even possible?

Reply:
I appreciate that the comment is lengthy. Perhaps it could be better worded.

`sourceURL` is _always_ present (let us ignore injected styles). It is typically the network resource url. If there's a magic sourceURL comment, it is instead the value of that comment. You can distinguish if `sourceURL` came from a comment if `hasSourceURL` is present.

canLink is an alias for "came from sourceURL", and is synonymous with "could it really be linked", which is the only concern in the details renderer, so I renamed it to `canLink` (as opposed to the more implementation-aware `cameFromSourceURLComment`).

Reply:
> I appreciate that the comment is lengthy. Perhaps it could be better worded.

hahaha I'm sorry @connorjclark I meant saga in terms of the whole process of font-size, not that your comment was a saga 😆 I greatly appreciate the detail 👍 

OK here's my stab at greatly reducing things combined with the above equality comments, let me know if it's right. If it is, I'm happy and feel free to adjust comments or not adjust comments as you see fit. If it's wrong, I'd love to actually understand what's going on :)

```
// DevTools protocol does not provide the resource URL if there was a magic `sourceURL` comment.
// `sourceURL` will be the raw value of the magic `sourceURL` comment which likely points to a file at build time, 
// not one that is served over the network we could resolve and link to.

Reply:
yeah that's better and cuts out the stuff that's not too important for this understanding this audit.